File: hello.kt
import kotlinx.cinterop.*
import libui.*

fun main(args: Array<String>) = memScoped {
    val options = alloc<uiInitOptions>()
    uiInit(options.ptr)

    val window = uiNewWindow("Konan говорит: click me!", 320, 60, 1)
    uiWindowSetMargined(window, 1)
    fun onClosing(window: CPointer<uiWindow>?, ptr: COpaquePointer?): Int {
        uiControlDestroy(window?.reinterpret())
        uiQuit()
        return 0
    }
    uiWindowOnClosing(window, staticCFunction(::onClosing), null)

    val box = uiNewVerticalBox()
    uiWindowSetChild(window, box?.reinterpret())

    val button = uiNewButton("Quit")
    fun onButtonClick(button: CPointer<uiButton>?, ptr: COpaquePointer?) {
        println("Hi Kotlin")
    }
    uiButtonOnClicked(button, staticCFunction(::onButtonClick), null)
    uiBoxAppend(box, button?.reinterpret(), 0)

    uiControlShow(window?.reinterpret())
    uiMain()
    uiUninit()
}

File: hello.c
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include "../../ui.h"

int onClosing(uiWindow *w, void *data)
{
	uiQuit();
	return 1;
}

void saySomething(uiButton *b, void *e)
{
	uiMultilineEntryAppend(uiMultilineEntry(e),
        "Hello, World!  Ciao, mondo!\n"
        "Привет, мир!  你好，世界！\n\n");
}

int main(void)
{
	uiInitOptions o;
	uiWindow *w;
	uiBox *b;
	uiButton *btn;
    uiMultilineEntry *e;

	memset(&o, 0, sizeof (uiInitOptions));
	if (uiInit(&o) != NULL)
		abort();

	w = uiNewWindow("Hello", 320, 240, 0);
	uiWindowSetMargined(w, 1);

	b = uiNewVerticalBox();
	uiBoxSetPadded(b, 1);
	uiWindowSetChild(w, uiControl(b));

	e = uiNewMultilineEntry();
	uiMultilineEntrySetReadOnly(e, 1);

	btn = uiNewButton("libui говорит: click me!");
	uiButtonOnClicked(btn, saySomething, e);
	uiBoxAppend(b, uiControl(btn), 0);

	uiBoxAppend(b, uiControl(e), 1);

	uiWindowOnClosing(w, onClosing, NULL);
	uiControlShow(uiControl(w));
	uiMain();
	return 0;
}
